core/storage/wal.rs

```diff
@@ -1108,4 +1364,557 @@ impl WalFileShared {
+        }
+
+        // After restart: max_frame == 0, nbackfills == 0, salts changed, seq incremented.
+        let (seq1, salt11, salt21, _ps2) = wal_header_snapshot(&wal_shared);
+        assert_eq!(seq1, seq0.wrapping_add(1), "checkpoint_seq increments");
+        assert_ne!(salt21, salt20);
+        assert_eq!(salt11, salt10.wrapping_add(1), "salt_1 should increment");
+        let (after_max, after_backfills) = unsafe {
+            let s = &*wal_shared.get();
+            (
+                s.max_frame.load(Ordering::SeqCst),
+                s.nbackfills.load(Ordering::SeqCst),
+            )
+        };
+        assert_eq!(after_max, 0);
+        assert_eq!(after_backfills, 0);
+
+        // Next write should create frame_id = 1 again.
+        conn.execute("insert into test(value) values ('post_restart')")
+            .unwrap();
+        conn.pager
+            .borrow_mut()
+            .wal
+            .borrow_mut()
+            .finish_append_frames_commit()
+            .unwrap();
+        let new_max = unsafe { (&*wal_shared.get()).max_frame.load(Ordering::SeqCst) };
+        assert_eq!(new_max, 1, "Sequence restarted at 1");
+
+        let meta_after = std::fs::metadata(&walpath).unwrap();
+        let bytes_after = meta_after.len();
+        assert_eq!(
+            bytes_before, bytes_after,
+            "WAL file should not change size after full checkpoint"
+        );
+        std::fs::remove_dir_all(path).unwrap();
+    }
+
+    #[test]
+    fn test_wal_passive_partial_then_complete() {
+        let (db, _tmp) = get_database();
+        let conn1 = db.connect().unwrap();
+        let conn2 = db.connect().unwrap();
+
+        conn1
+            .execute("create table test(id integer primary key, value text)")
+            .unwrap();
+        bulk_inserts(&conn1.clone(), 15, 2);
+        conn1.pager.borrow_mut().cacheflush().unwrap();
+
+        // Force a read transaction that will freeze a lower read mark
+        {
+            let pager = conn2.pager.borrow_mut();
+            let mut wal2 = pager.wal.borrow_mut();
+            assert!(matches!(wal2.begin_read_tx().unwrap(), LimboResult::Ok));
+        }
+
+        // generate more frames that the reader will not see.
+        bulk_inserts(&conn1.clone(), 15, 2);
+        conn1.pager.borrow_mut().cacheflush().unwrap();
+
+        // Run passive checkpoint, expect partial
+        let (res1, max_before) = {
+            let pager = conn1.pager.borrow();
+            let mut wal = pager.wal.borrow_mut();
+            let res = run_checkpoint_until_done(&mut *wal, &pager, CheckpointMode::Passive);
+            let maxf = unsafe {
+                (&*db.maybe_shared_wal.read().as_ref().unwrap().get())
+                    .max_frame
+                    .load(Ordering::SeqCst)
+            };
+            (res, maxf)
+        };
+        assert_eq!(res1.num_wal_frames, max_before);
+        assert!(
+            res1.num_checkpointed_frames < res1.num_wal_frames,
```

Comment:
Can we make this assertion more precise (i.e. we know we only checkpointed up to reader's mark)?